crates/eth-rpc/src/main.rs

```diff
@@ -16,17 +17,61 @@ async fn main() -> Result<()> {
     tracing_subscriber::FmtSubscriber::builder().with_env_filter(filter).finish().try_init()?;
 
     let starknet_config = StarknetConfig::from_env()?;
+
     let rpc_config = RPCConfig::from_env()?;
 
-    let provider = JsonRpcClientBuilder::with_http(&starknet_config).unwrap().build();
-    let kakarot_client = KakarotClient::new(starknet_config, provider);
+    // This match creates redundancy but it appears KakarotClient<P> complains about the
+    // the possibility of P being either a JsonRpcClient or a SequencerGatewayProvider, but not being
+    // known at compile time.
+    match starknet_config.network {
+        Network::Madara | Network::Katana => {
+            let provider = JsonRpcClientBuilder::with_http(&starknet_config).unwrap().build();
+            let kakarot_client = KakarotClient::new(starknet_config, provider);
+
+            let (server_addr, server_handle) = run_server(Box::new(kakarot_client), rpc_config).await?;
+            let url = format!("http://{server_addr}");
+
+            println!("RPC Server running on {url}...");
+
+            server_handle.stopped().await;
+        }
+        Network::Goerli1 => {
+            let provider = SequencerGatewayProvider::starknet_alpha_goerli();
+            let kakarot_client = KakarotClient::new(starknet_config, provider);
+
+            let (server_addr, server_handle) = run_server(Box::new(kakarot_client), rpc_config).await?;
+            let url = format!("http://{server_addr}");
+
+            println!("RPC Server running on {url}...");
+
+            server_handle.stopped().await;
+        }
+        Network::Goerli2 => {
+            let provider = SequencerGatewayProvider::starknet_alpha_goerli_2();
+            let kakarot_client = KakarotClient::new(starknet_config, provider);
+
+            let (server_addr, server_handle) = run_server(Box::new(kakarot_client), rpc_config).await?;
+            let url = format!("http://{server_addr}");
+
+            println!("RPC Server running on {url}...");
+
+            server_handle.stopped().await;
+        }
+        Network::Mainnet => {
+            let provider = SequencerGatewayProvider::starknet_alpha_mainnet();
+            let kakarot_client = KakarotClient::new(starknet_config, provider);
 
-    let (server_addr, server_handle) = run_server(Box::new(kakarot_client), rpc_config).await?;
-    let url = format!("http://{server_addr}");
+            let (server_addr, server_handle) = run_server(Box::new(kakarot_client), rpc_config).await?;
+            let url = format!("http://{server_addr}");
 
-    println!("RPC Server running on {url}...");
+            println!("RPC Server running on {url}...");
 
-    server_handle.stopped().await;
+            server_handle.stopped().await;
+        }
+        Network::Mock => {
+            return Err(eyre::eyre!("Cannot run RPC server with mock network"));
+        }
+    }
```

Comment:
There seems to be some redundency, maybe we could only initialize the provider within the match and kakarot_client, server_addr, server_handle and url after the match

Reply:
I tried but didn't manage to.
KakarotClient<P> complains when the P is not known at compile time. In the match arms, there could be a JsonRpcClient or a SequencerGatewayProvider.